jetty-jmx/src/main/java/org/eclipse/jetty/jmx/MBeanContainer.java

```diff
@@ -122,6 +114,215 @@ public String getDomain()
+     */
+    public Object mbeanFor(Object o)
+    {
+        if (o==null)
+            return null;
+
+        Constructor<?> constructor = findMBeanFor(o.getClass());
+
+        if (constructor==null)
+            return null;
+
+        try
+        {
+            Object mbean;
+            if (constructor.getParameterCount()==0)
+            {
+                mbean = constructor.newInstance();
+                ((ModelMBean)mbean).setManagedResource(o, "objectReference");
+            }
+            else
+            {
+                mbean = constructor.newInstance(o);
+            }
+
+            if (mbean instanceof ObjectMBean)
+                ((ObjectMBean)mbean).setMBeanContainer(this);
+
+            if (LOG.isDebugEnabled())
+                LOG.debug("mbeanFor {} is {}", o, mbean);
+
+            return mbean;
+        }
+        catch(Throwable t)
+        {
+            LOG.warn(t);
+        }
+
+        return null;
+    }
+
+    private Constructor<?> findMBeanFor(Class<?> oClass)
+    {
+        if (oClass==null)
+            return null;
+
+        Constructor<?> constructor = _mbeanFor.get(oClass);
+        if (constructor!=null)
+            return constructor;
+
+        try
+        {
+            String pName = oClass.getPackage().getName();
+            String cName = oClass.getName().substring(pName.length() + 1);
+            String mName = pName + ".jmx." + cName + "MBean";
+
+            Class<?> mClass;
+            try
+            {
+                // Look for an MBean class from the same loader that loaded the original class
+                mClass = (Object.class.equals(oClass))?oClass = ObjectMBean.class:Loader.loadClass(oClass, mName);
+            }
+            catch (ClassNotFoundException e)
+            {
+                // Not found, so if not the same as the thread context loader, try that.
+                if (Thread.currentThread().getContextClassLoader() == oClass.getClassLoader())
+                    throw e;
+                LOG.ignore(e);
+                mClass = Loader.loadClass(oClass, mName);
+            }
+
+            constructor =  (ModelMBean.class.isAssignableFrom(mClass))
+                ?mClass.getDeclaredConstructor()
+                :mClass.getConstructor(OBJ_ARG);
+        }
+        catch (ClassNotFoundException | NoSuchMethodException e)
+        {
+            LOG.debug(e.toString());
+            LOG.ignore(e);
+        }
+
+        if (constructor==null)
+            constructor = findMBeanFor(oClass.getSuperclass());
+
+        // Can we cache the result?
+        ClassLoader ourLoader = this.getClass().getClassLoader();
+        if (constructor!=null && (_cacheOtherClassLoaders || oClass.getClassLoader()==ourLoader && constructor.getDeclaringClass().getClassLoader()==ourLoader))
+        {
+            Constructor<?> c = _mbeanFor.putIfAbsent(oClass, constructor);
+            if (c!=null)
+                constructor = c;
+        }
```

Comment:
I think we should also cache the negative result. We have made all the effort to figure out whether e.g. `HTTP2Session` has an MBean; we don't want to do all the work again for the 2nd instance of `HTTP2Session` that we're creating.
Can't we use a constructor constant to put in the map, and if we find it, return `null`?

Reply:
There should never be a negative result.   All classes will eventually go to `Object`, which will be mapped to `ObjectMBean`.  So null is not an option.